src/impls/serde.rs

```diff
@@ -0,0 +1,105 @@
+use crate::U256;
+use alloc::string::String;
+use core::{fmt, str::FromStr};
+use crate_serde::{
+    de::{self, Deserialize, Deserializer},
+    ser::{Serialize, Serializer},
+};
+
+impl Serialize for U256 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        static CHARS: &[u8] = b"0123456789abcdef";
+
+        fn to_hex_raw<'a>(v: &'a mut [u8], bytes: &[u8]) -> &'a str {
+            assert!(v.len() > 1 + bytes.len() * 2);
+
+            v[0] = b'0';
+            v[1] = b'x';
+
+            let mut idx = 2;
+            let first_nibble = bytes[0] >> 4;
+            if first_nibble != 0 {
+                v[idx] = CHARS[first_nibble as usize];
+                idx += 1;
+            }
+            v[idx] = CHARS[(bytes[0] & 0xf) as usize];
+            idx += 1;
+
+            for &byte in bytes.iter().skip(1) {
+                v[idx] = CHARS[(byte >> 4) as usize];
+                v[idx + 1] = CHARS[(byte & 0xf) as usize];
+                idx += 2;
+            }
+
+            // SAFETY: all characters come either from CHARS or "0x", therefore valid UTF8
+            unsafe { core::str::from_utf8_unchecked(&v[0..idx]) }
+        }
+
+        let mut slice = [0u8; 66];
+        let bytes = self.to_be_bytes();
+
+        let non_zero = bytes.iter().take_while(|&&b| b == 0).count();
+        let bytes = &bytes[non_zero..];
+        if bytes.is_empty() {
+            serializer.serialize_str("0x0")
+        } else {
+            serializer.serialize_str(to_hex_raw(&mut slice, bytes))
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for U256 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl<'a> de::Visitor<'a> for Visitor {
+            type Value = U256;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                write!(formatter, "a (both 0x-prefixed or not) hex string",)
+            }
+
+            fn visit_str<E: de::Error>(self, v: &str) -> Result<Self::Value, E> {
+                Ok(if let Some(v) = v.strip_prefix("0x") {
+                    U256::from_str_radix(v, 16).map_err(de::Error::custom)?
+                } else {
+                    U256::from_str(v).map_err(de::Error::custom)?
```

Comment:
What is the reasoning to allow permissive deserialization here? I know that the Ethereum JSON RPC standard has very strict rules in that:
- Requires a `0x` prefix
- No leading 0's allowed.

Should that be the default `deserialize` behaviour (seeing as it mirrors the `serialize` implementation).

Reply:
When people use Foundry's `cast`, they sometimes will write `cast send vitalik.eth 123` which is a decimal number, and requires us to manually check if it starts with 0x, to deserialize as decimal, else as hex. This manfiests in many places in the codebase, whereas having it in the lib would help